src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,215 @@
+#include <app/AttributeAccessInterface.h>
+#include <app/AttributePersistenceProvider.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/util/af.h>
+#include <map>
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace ModeBase {
+
+class Instance : public CommandHandlerInterface, public AttributeAccessInterface
+{
+public:
+    /**
+     * This is a helper function to build a mode option structure. It takes the label/name of the mode,
+     * the value of the mode and a list of mode tags that apply to this mode.
+     *
+     * **NOTE** The caller must ensure that the lifetime of the label and modeTags is as long as the
+     * returned structure.
+     */
+    static chip::app::Clusters::detail::Structs::ModeOptionStruct::Type
+    BuildModeOptionStruct(const char * label, uint8_t mode,
+                          const DataModel::List<const chip::app::Clusters::detail::Structs::ModeTagStruct::Type> modeTags)
+    {
+        chip::app::Clusters::detail::Structs::ModeOptionStruct::Type option;
+        option.label    = CharSpan::fromCharString(label);
+        option.mode     = mode;
+        option.modeTags = modeTags;
+        return option;
+    }
+
+    /**
+     * Initialise the ModeBase server instance.
+     * @return Returns an error if the cluster ID given is not af a valid ModeBase cluster, if the
+     * given endpoint and cluster ID have not been enabled in zap, if the CommandHandler or
+     * AttributeHandler registration fails or if the AppInit() returns an error.
+     */
+    CHIP_ERROR Init();
+
+    // CommandHandlerInterface
+    void InvokeCommand(HandlerContext & ctx) override;
+    CHIP_ERROR EnumerateAcceptedCommands(const ConcreteClusterPath & cluster, CommandIdCallback callback, void * context) override;
+
+    // AttributeAccessInterface
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder) override;
+
+    // Attribute setters
+    chip::Protocols::InteractionModel::Status UpdateStartUpMode(DataModel::Nullable<uint8_t> aNewStartUpMode);
+    chip::Protocols::InteractionModel::Status UpdateOnMode(DataModel::Nullable<uint8_t> aNewOnMode);
+    chip::Protocols::InteractionModel::Status UpdateCurrentMode(uint8_t aNewMode);
+
+    // Attribute getters
+    DataModel::Nullable<uint8_t> GetStartUpMode() const;
+    DataModel::Nullable<uint8_t> GetOnMode() const;
+    uint8_t GetCurrentMode() const;
+    EndpointId GetEndpointId() const { return mEndpointId; }
+
+    /**
+     * Returns true if the feature is supported.
+     * @param feature the feature to check.
+     */
+    bool HasFeature(Feature feature) const;
+
+    /**
+     * This function returns true if the mode value given matches one of the supported modes, otherwise it returns false.
+     * @param mode
+     */
+    bool IsSupportedMode(uint8_t mode);
+
+private:
```

Comment:
We should put all the "public and to be overridden" API (what consumers need to know) together, without implementation details interleaved with it.  All the private stuff should come after it, or before it, or something, but not in between.